node/pegnet/grading.go

```diff
@@ -50,22 +50,53 @@ const createTableRate = `CREATE TABLE IF NOT EXISTS "pn_rate" (
 );
 `
 
-func (p *Pegnet) InsertRate(tx *sql.Tx, height uint32, rates []opr.AssetUint) error {
+func (p *Pegnet) insertRate(tx *sql.Tx, height uint32, ticker fat2.PTicker, rate uint64) error {
+	_, err := tx.Exec("INSERT INTO pn_rate (height, token, value) VALUES ($1, $2, $3)", height, ticker.String(), rate)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// InsertRates adds all asset rates as rows, computing the rate for PEG if necessary
+func (p *Pegnet) InsertRates(tx *sql.Tx, height uint32, rates []opr.AssetUint, pricePEG bool) error {
 	for _, r := range rates {
-		// TODO: Make this more robust? Check if this is the best place to do this
-		if r.Name != "PEG" {
-			r.Name = "p" + r.Name
+		if r.Name == "PEG" {
+			continue
+		}
+		r.Name = "p" + r.Name
+		err := p.insertRate(tx, height, fat2.StringToTicker(r.Name), r.Value)
+		if err != nil {
+			return err
 		}
-		_, err := tx.Exec("INSERT INTO pn_rate (height, token, value) VALUES ($1, $2, $3)", height, r.Name, r.Value)
+	}
+	var ratePEG uint64
+	if pricePEG {
+		// PEG price = (total capitalization of all other assets) / (total supply of all other assets at height - 1)
+		issuance, err := p.SelectIssuances()
 		if err != nil {
 			return err
 		}
+		var totalIssuance uint64
+		var totalCapitalization uint64
+		for _, r := range rates {
+			if r.Name == "PEG" {
+				continue
+			}
+			assetIssuance := issuance[fat2.StringToTicker(r.Name)]
+			totalIssuance += assetIssuance
+			totalCapitalization += assetIssuance * r.Value
+		}
+		ratePEG = totalCapitalization / totalIssuance
```

Comment:
This is incorrect. It is, `Total Market Cap of all Assets / PEG Supply`

You can see how I implemented it here: https://github.com/pegnet/pegnet/blob/PNT-328-peg_source/polling/pegsource.go#L67-L98

You want to do a range over all the assets, and total up their `Value * Issuance`. You will want to use a BigInt for this.
Then you divide by the PEG supply (https://github.com/pegnet/pegnet/blob/PNT-328-peg_source/polling/pegsource.go#L97). I check for 0, as a 0 PEG system could happen in local nets for example.

Reply:
Gotcha yeah had a brain fart and divided by the total supplies rather than the PEG supply, which makes more sense. Was going to double check myself in the morning but you did the hard work for me :)

I'll switch the total capitalization to a big int though

Reply:
But also where you check for the zero value of PEG supply, it should just return the total capitalization instead of 0. As then it will be impossible for someone to convert into PEG raising the supply again

Reply:
I figure PEG will come from inflation. But that is an interesting point, I think that failure mode would be fine. It should realistically never happen on mainnet